pkg/suggestion/genetic_algorithm/genetic_algorithm.go

```diff
@@ -0,0 +1,313 @@
+	selection                    string  // either roulette or elite; recommend elite
+	selectNum                    int     // must be >=2; recommend 2 or 3
+	crossover                    string  // only uniform
+	geneMutation                 string  // either random or perturbation; recommend random
+	generationChange             string  // either discrete or continuous
+	evaluationFunction           string  // evaluation function
+	evaluateHigh                 bool    // true for higher is better; false for lower is better
+	geneNames                    []string
+	geneTypeMap                  GeneTypeMap
+}
+
+func NewGA(numGenes int, numOffsprings int, geneMutationProbability float64, offspringMutationProbability float64, maxGenerations int, selection string, selectNum int, crossover string, geneMutation string, generationChange string, evaluationFunction string, evaluateHigh bool, geneNames []string, geneTypeMap GeneTypeMap) *GA {
+	return &GA{numGenes, numOffsprings, geneMutationProbability, offspringMutationProbability, maxGenerations, selection, selectNum, crossover, geneMutation, generationChange, evaluationFunction, evaluateHigh, geneNames, geneTypeMap}
+}
+
+// randomly generate a generation; for the initial generation to start with
+func (ga *GA) InitialRandomSuggest(geneTypeMap GeneTypeMap, genes Genes) Generation {
+	offsprings := []Offspring{}
+	for i := 0; i < ga.numOffsprings; i++ {
+		genes.setRandomGeneValues(geneTypeMap)
+		offspring := NewOffspring(genes)
+		offsprings = append(offsprings, *offspring)
+	}
+	generation := NewGeneration(0)
+	generation.setGeneration(offsprings)
+	return *generation
+}
+
+// genetic algorithm optimizer; optimizes against evaluateFunc with GA variables
+func (ga *GA) Optimize(evaluateFunc func(offspring Offspring) float64) (Offspring, float64) {
+	genes := NewGenes(ga.geneNames)
+	bestOffspring := *NewOffspring(genes)
+	var bestScore float64
+
+	// initial generation and the scores of its offsprings
+	generation := ga.InitialRandomSuggest(ga.geneTypeMap, genes)
+	gaResult := NewGAResult()
+	for i, o := range generation.offsprings {
+		gaResult.recordResult(i, evaluateFunc(o))
+	}
+	bestOffsprings := ga.Select(*gaResult, generation)
+
+	// run genetic algorithm for maxGeneration iterations to optimize genes
+	for g := 0; g < ga.maxGenerations; g++ {
+		generation = ga.Crossover(bestOffsprings, generation)
+		gaResult = NewGAResult()
+		for i, o := range generation.offsprings {
+			gaResult.recordResult(i, evaluateFunc(o))
+		}
+		bestOffsprings = ga.Select(*gaResult, generation)
+	}
+
+	// best genes after optimization
+	bestScoreMap := ga.GetBestScores(*gaResult, 1)
+	for k, v := range bestScoreMap {
+		bestScore = v
+		bestOffspring = generation.offsprings[k]
+	}
+
+	// returns the best optimized offspring and its score
+	return bestOffspring, bestScore
+}
+
+// a set of evaluation result for each offspring
+type GAResult struct {
+	result map[int]float64
```

Comment:
I think we should use a list of a struct that has a Sorter interface.
At `GetBestScores` you can sort and get a slice.
It is better for both performance and readable.

Reply:
Oh, that looks better. I'll make it this weekend.